url/src/main/java/com/datastax/oss/dsbulk/url/S3URLStreamHandler.java

```diff
@@ -129,16 +138,33 @@ S3Client getS3Client(String query) {
       return builder.build();
     }
 
-    private Map<String, String> parseParameters(String query) {
-      Map<String, String> parameters = new HashMap<>();
-      String[] paramList = query.split("&", 0);
-      for (String param : paramList) {
-        String[] parts = param.split("=", 0);
-        if (parts.length > 1 && !StringUtils.isBlank(parts[1])) {
-          parameters.put(parts[0], parts[1]);
+    // Borrowed from
+    // https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection
+    private static Map<String, List<String>> parseParameters(String query)
+        throws UnsupportedEncodingException {
+      final Map<String, List<String>> queryPairs = new LinkedHashMap<>();
+      final String[] pairs = query.split("&", 0);
+      for (String pair : pairs) {
+        final int idx = pair.indexOf("=");
+        final String key = idx > 0 ? URLDecoder.decode(pair.substring(0, idx), "UTF-8") : pair;
+        queryPairs.computeIfAbsent(key, k -> new ArrayList<>());
+        final String value =
+            idx > 0 && pair.length() > idx + 1
+                ? URLDecoder.decode(pair.substring(idx + 1), "UTF-8")
+                : null;
+        queryPairs.get(key).add(value);
+      }
+      return queryPairs;
+    }
+
+    private static String getQueryParam(Map<String, List<String>> parameters, String key) {
+      if (parameters.containsKey(key)) {
+        List<String> values = parameters.get(key);
+        if (values != null && !values.isEmpty()) {
+          return values.get(0);
```

Comment:
We are only considering the first occurrence of each parameter, and I think that's indeed enough. But in this case maybe we should simplify `parseParameters` to return `Map<String,String>`, wdyt?

Reply:
I have no strong preferences either way. :)

Reply:
OK let's leave it as is.